### ISSUE TITLE
risk brought by  no pre-allocate output for allgather&reducescatter&alltoall

### ISSUE THREAD

**Issue by mingxiaoh** [horovod/horovod#3931]
Hello, recently I was trying to enable horovod torch UT on xpu backend, and then I noticed one issue, torch allgather&reducescatter UT will fail randomly when hvd.rank > 4,  if hvd.rank is bigger, the failure possibility will be bigger. And then I noticed one issue, for https://github.com/horovod/horovod/blob/master/horovod/torch/mpi_ops.py#L626, seems that you don't pre-allocate output for allgather(similar to reducescatter&alltoall), in this case, it will be conflict with this rule "Contexts and controller must be initialized and the background thread must be running before this function is called. " 
https://github.com/horovod/horovod/blob/master/horovod/common/operations.cc#L1581-L1582

Given this, I tried to pre-allocate output for allgather&reducescatter, e.g., change  https://github.com/horovod/horovod/blob/master/horovod/torch/mpi_ops.py#L626 from "output = tensor.new()" to "output = tensor.new(tensor.shape)", I found no random failure after this change. 

I would like to try it on cuda but I have no such more cuda at hand now, may you please help explain the logic? Thanks.

**Comment by mingxiaoh** [comment]
@maxhgerlach Thanks for the detailed reply.
our test case is like "mpirun -np 10 -ppn 10 pytest -v -k "allgather"  test_torch.py", take  test_horovod_allgather for example, 
the value of https://github.com/horovod/horovod/blob/master/test/parallel/test_torch.py#L1113 is wrong, leading to 
https://github.com/horovod/horovod/blob/master/test/parallel/test_torch.py#L1122-L1123 raise failure.
error will be missing if we change https://github.com/horovod/horovod/blob/master/horovod/torch/mpi_ops.py#L626 from "output = tensor.new()" to "output = tensor.new(tensor.shape)" or "output = tensor.new(1)", besides, error will also be missing if we just add hvd.barrier() after this line(https://github.com/horovod/horovod/blob/master/test/parallel/test_torch.py#L1113). 

as for reducescatter, yes, we can still see random error based on v0.28, but the reproducing possibility is quite slow(about 10%). So, pre-allocate the output before enqueuing might not be a fix for this issue.

Finally, according to my colleague, he said horovod tensoflow meet the same issue, and he tried to follow my idea to do pre-allocate for allgather, which is similar to allreduce here(https://github.com/horovod/horovod/blob/master/horovod/tensorflow/mpi_ops.cc#L488), problem is fixed.

**Comment by maxhgerlach** [comment]
> trying to enable horovod torch UT on xpu backend

I still haven't understood the context of this report. Is this related to https://github.com/intel/intel-extension-for-pytorch/ ? What is the meaning of "torch UT"? I do not know the abbreviation.

Currently, Horovod has code to deal with asynchronous memory allocations on GPUs (CUDA and -- to a certain extend -- ROCm), while CPU memory allocations are assumed to be synchronous. It sounds like you may be running into race conditions because such a distinction is not done correctly for your backend. I'm not sure your proposed workaround fully addresses that.

**Comment by mingxiaoh** [comment]
@maxhgerlach  sorry for late feedback, I have been checking this issue. 
torch UT means this file https://github.com/horovod/horovod/blob/master/test/parallel/test_torch.py, 
we know that Horovod has code to deal with asynchronous memory allocations on GPUs(CUDA, XPU), but hvd.allgather() for torch (details see https://github.com/horovod/horovod/blob/master/horovod/torch/mpi_ops.py#L637-L638I ) is synchronous on GPUs or just on CPU only? If just on CPU only, shouldn't we add hvd.barrier() after this line https://github.com/horovod/horovod/blob/master/test/parallel/test_torch.py#L1113?


BTW, as for your question regarding reproduction on CPU and cuda, 
first of all, we could not reproduce it with CPU path, secondly, we can reproduce it randomly with xpu backend(implemented in https://github.com/intel/intel-optimization-for-horovod) when hvd.size>=10, thirdly, we have no so many cuda devices at hand, so we have no chance to try cuda.

**Comment by maxhgerlach** [comment]
Thanks for the explanations of the context of this issue, @mingxiaoh! 

> error will be missing if we change https://github.com/horovod/horovod/blob/master/horovod/torch/mpi_ops.py#L626 from "output = tensor.new()" to "output = tensor.new(tensor.shape)" or "output = tensor.new(1)", besides, error will also be missing if we just add hvd.barrier() after this line(https://github.com/horovod/horovod/blob/master/test/parallel/test_torch.py#L1113).

While the `tensor.new(1)` seems to mask the problem, maybe by introducing a small delay, it would be better to understand the underlying cause. Else it might introduce data corruption in production work loads, even though it's not been seen in unit tests. Likewise, `hvd.barrier()` would be a slow band-aid. It's a global synchronization point that blocks all distributed Horovod processes.

Earlier, there was a race condition with output allocations in Horovod's PyTorch backend  because CPU-GPU synchronization was not done as needed (fix in https://github.com/horovod/horovod/pull/3639). Could there be a similar problem that is specific to the fork for Intel GPUs?

**Comment by mingxiaoh** [comment]
@maxhgerlach  our code is based on v0.28, so, fix in (https://github.com/horovod/horovod/pull/3639) has already been in our code base.

**Comment by romerojosh** [comment]
I think our issue here is that the proposed fix of adding `tensor.new(1)` to create a small preallocated array does not actually solve a possible race condition on the output allocation for allgather and similar operations. Even if you preallocate a small output tensor like this, it still needs to be resized when the allgather operation executes and the actual output size is known, which triggers a separate allocation + copy operation in PyTorch. If the issue you are suspecting arises due a race condition on the output allocation, the issue would still persist.

For that reason, it'd be good to get a more complete explanation of the issue you are running into so we can help identify a real fix (whether it be in this Horovod repository or the Intel GPU fork you are using). There may be some assumption built into our code that works for all our currently tested platforms but breaks down in your use case.

**Comment by mingxiaoh** [comment]
@romerojosh  thanks, we found some hints in oneCCL but need further debug, will feedback once we have progress.
